Fix race conditions in KafkaStreams startup

pcasaes opened the pull request:
Fixes #10724

pcasaes commented:
Fixes #10724

gsmet commented:
@gunnarmorling @mkouba could you have a look at that one?

mkouba commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -189,7 +191,7 @@ private static String asString(List<InetSocketAddress> addresses) {
     }
 
-    void onStart(@Observes StartupEvent ev) {
+    void onStart(@Observes @Priority(Interceptor.Priority.LIBRARY_BEFORE + 500) StartupEvent ev) {
```
I have no idea what this code is doing but this observer will be notified _before_ similar observers with no priority declared (the default priority is 2500, `Priority.LIBRARY_BEFORE + 500 = 1500` and observers with smaller priority values are called first).

If that's the intention I have nothing to add ;-).

gsmet replied:
Yeah, I should have been clearer :).

I'm not totally sure this pattern of doing things with startup events and then returning what's initialized here is a good thing. I'm pretty sure this could lead to weird startup issues.

But I'm not a Kafka Streams expect, maybe it's how they do things to avoid waiting for things to be fully started.

gunnarmorling replied:
Is this issue specific to Kafka Streams even, or is it something more generic? I'd have expected that the dependency resolver of CDI takes care of this: i.e. if there's an injection point for `KafkaStreams` (or any type, really), it would make sure the corresponding producer is started up before that?

Also it's interesting that an NPE happens, I'd rather have expected some sort of "UnresolvedDependencyInjection"?

mkouba replied:
@gunnarmorling, dependency injection and CDI events are two different concepts. If there is an injection point of type `KafkaStreams` then CDI ensures that the corresponding bean is created before the injection point is injected.

> Also it's interesting that an NPE happens, I'd rather have expected some sort of "UnresolvedDependencyInjection"?

`UnsatisfiedResolutionException` is thrown if there is an injection point that cannot be satisfied. Which is not the case here  - the producer method would satisfy such an injection point. However, the `KafkaStreamsTopologyManager#streams` could be _uninitialized_ and thus the producer method returns null.

So the real problem is - when/how to initalize the streams so that a bean may not inject an uninitialized producer method.

This morning I sent a PR that describes this "late init" problem with extensions:
https://github.com/quarkusio/quarkus/blob/f12c24e4be88fd3d7d077ed9087fe8ffecc83f56/docs/src/main/asciidoc/writing-extensions.adoc#user-content-bean_init

Unfortunately, I'm no kafka streams expert either so I can't answer your question @gsmet...

gunnarmorling replied:
> the producer method would satisfy such an injection point. However, the KafkaStreamsTopologyManager#streams could be uninitialized and thus the producer method returns null.

I see. I think this code is based on the assumpt that if there's a producer method like in this case here, the @Observes StartupEvent method of the bean declaring that method is guaranteed to have been run before that producer method is invoked. Seems like that assumption doesn't hold as per what your say.

In that light I'm wondering why we don't start the KafkaStreams object simply in the constructor instead of using the start-up observer method.

mkouba replied:
> StartupEvent method of the bean declaring that method is guaranteed to have been run before that producer method is invoked. Seems like that assumption doesn't hold as per what your say.

Yes, completely unfounded assumption.

> In that light I'm wondering why we don't start the KafkaStreams object simply in the constructor instead of using the start-up observer method.

Maybe it has to be started at runtime and some other bean needs to access `KafkaStreamsTopologyManager` during static init?

rquinio replied:
I think the problem is `KafkaStreamsTopologyManager#setRuntimeConfig` called by KafkaStreamsRecorder has to happen before KStream can be started (because once a KStreams instance is started it can't be re-configured afterward). And I guess KafkaStreamsRuntimeConfig is not a bean that could be directly injected in the constructor either ?

mkouba replied:
`KafkaStreamsRuntimeConfig` can be injected in a bean. However, it is a RUN_TIME config root so it may not be available before the app starts. E.g. if the consuming bean is created during STATIC_INIT you will get `CreationException`.

gunnarmorling replied:
> Yes, completely unfounded assumption.

Well, I'd say it's not so far-fetched to assume a bean would be fully initialized and ready to use before any of its methods are invoked by the runtime. But ok, things are as they are.

> Maybe it has to be started at runtime

Yes, it absolutely must be started at runtime. `KafkaStreamsTopologyManager` shouldn't be needed at static init IIRC. So can we have it being instantiated at runtime only? Or maybe the suggested priority fix is just fine really :)

mkouba replied:
>> Yes, completely unfounded assumption.

> Well, I'd say it's not so far-fetched to assume a bean would be fully initialized and ready to use before any of its methods are invoked by the runtime. But ok, things are as they are.

@gunnarmorling The bean itself is fully initialized but its state is not initialized. In this particular case there is a bean that declares a producer method and an obsever. The producer method is backed by a state that is _explicitly initialized_ after the bean is created and has no connection to the observer except that they both are invoked upon the same bean instance. In other words, it is only guaranteed that the `KafkaStreamsTopologyManager` is created before the producer or the observer is called.

pcasaes replied:
Just to add. I did debug the code and it is showing non deterministic behavior. Sometimes `KafkaStreamsTopologyManager` would be initialized before the bean expecting the `KafkaStreams` and other times not. 

```
    @Inject
    KafkaStreams kafkaStreams;

    void startUp(@Observes StartupEvent ev) {
        LOG.info("Initializing metrics exporter");
    }

    @PostConstruct
    void init() {

        LOG.info("My stream as a String: " + kafkaStreams);
        // Sometimes it gives a NPE
        var metrics = kafkaStreams.metrics();

    }
```

A quick work around was to change the priority to

```
    void startUp(@Observes @Priority(Interceptor.Priority.APPLICATION + 600) StartupEvent ev) {
        LOG.info("Initializing metrics exporter");
    }
```

Which is what brought me to the code in this PR.

mkouba replied:
>  I did debug the code and it is showing non deterministic behavior.

Yes, that's expected. Your `StartupEvent` observer has also the default priority - if multiple observers define the same priority the ordering is undefined (random).

gsmet replied:
@mkouba can't we just move whatever is in the startup event in the `@Produces` method?

mkouba replied:
Well, that could be a solution. However, it could still break things if someone attempts to access the produced bean during STATIC_INIT....  

But again I have no idea about the "business logic" of this bean...

vietk replied:
I think it won't work in all cases, because it's not necessary that an application injects explicitly a KafkaStreams instance into another bean, more over if it is injected in an @applicationScoped bean, it will be created after the application has started, not sure if it's the expected behavior. 

Maybe the best would be to change the scope of the Kafkastreams method to @applicationScoped (it's currently a Singleton) and create a holder class to be sure to access the kafkastreams at runtime.

gunnarmorling replied:
> However, it could still break things if someone attempts to access the produced bean during STATIC_INIT...

The `KafkaStreams` object is only meant to be used at runtime. It processes data from Kafka topics. You could compare it to creating a database connection which also doesn't make sense at static init.

mkouba replied:
> The KafkaStreams object is only meant to be used at runtime.

In that case, we should probably throw an exception if someone attempts to use the bean during STATIC_INIT...

gsmet replied:
> In that case, we should probably throw an exception if someone attempts to use the bean during STATIC_INIT...

Well, we don't really do that for all the other beans that shouldn't be accessed at static init, do we? At least I never took care of this.

If moving things to the `@Produces` method is OK, that would have my preference. (And I would like to include it in 1.6.1)

pcasaes replied:
Just pushed the change. Moved the instantiation to the producer method. As a consequence it no longer will be eager loaded though. Also will throw an exception if the Topology is not defined, not sure if this is the right approach.

gunnarmorling commented:
> the producer method would satisfy such an injection point. However, the KafkaStreamsTopologyManager#streams could be uninitialized and thus the producer method returns null.

I see. I think this code is based on the assumpt that if there's a producer method like in this case here, the `@Observes StartupEvent` method of the bean declaring that method is guaranteed to have been run before that producer method is invoked. Seems like that assumption doesn't hold as per what your say.

In that light I'm wondering why we don't start the `KafkaStreams` object simply in the constructor instead of using the start-up observer method.

pcasaes committed:
Instantiate KafkaStreams in producer method

rquinio commented:
Just checked the CI, there's a failure in quarkus-integration-test-kafka-streams native tests:

```
2020-07-16T18:14:58.9502207Z 2020-07-16 18:14:58,040 ERROR [io.sma.health] (vert.x-worker-thread-0) SRHCK01000: Error processing Health Checks: java.lang.NullPointerException
2020-07-16T18:14:58.9502903Z 	at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:103)
2020-07-16T18:14:58.9503270Z 	at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:142)
2020-07-16T18:14:58.9503870Z 	at org.apache.kafka.clients.admin.AdminClientConfig.<init>(AdminClientConfig.java:216)
2020-07-16T18:14:58.9504923Z 	at org.apache.kafka.clients.admin.Admin.create(Admin.java:61)
2020-07-16T18:14:58.9506031Z 	at org.apache.kafka.clients.admin.AdminClient.create(AdminClient.java:39)
2020-07-16T18:14:58.9508248Z 	at io.quarkus.kafka.streams.runtime.KafkaStreamsTopologyManager.getMissingTopics(KafkaStreamsTopologyManager.java:281)
2020-07-16T18:14:58.9510077Z 	at io.quarkus.kafka.streams.runtime.KafkaStreamsTopologyManager_ClientProxy.getMissingTopics(KafkaStreamsTopologyManager_ClientProxy.zig:308)
2020-07-16T18:14:58.9516388Z 	at io.quarkus.kafka.streams.runtime.health.KafkaStreamsTopicsHealthCheck.call(KafkaStreamsTopicsHealthCheck.java:50)
2020-07-16T18:14:58.9518406Z 	at io.quarkus.kafka.streams.runtime.health.KafkaStreamsTopicsHealthCheck_ClientProxy.call(KafkaStreamsTopicsHealthCheck_ClientProxy.zig:154)
(...)
at io.quarkus.smallrye.health.runtime.SmallRyeHealthHandlerBase.handle(SmallRyeHealthHandlerBase.java:17)
2020-07-16T18:14:58.9642219Z 	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
2020-07-16T18:14:58.9643059Z 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
2020-07-16T18:14:58.9644295Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-07-16T18:14:58.9645518Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-07-16T18:14:58.9657124Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2020-07-16T18:14:58.9657689Z 	at java.lang.Thread.run(Thread.java:834)
2020-07-16T18:14:58.9668525Z 	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
2020-07-16T18:14:58.9670109Z 	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
```

Btw that's a rare issue that was actually already happening today (got it once in JVM mode), if a readiness check gets triggered before the CDI startup event (race condition with Vert.x router startup I guess).

A quick solution could be to inject KafkaStreams inside KafkaStreamsTopicsHealthCheck (even if not used) to make sure KafkaStreamsTopologyManager is full initialized ?

gsmet commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -189,56 +188,56 @@ private static String asString(List<InetSocketAddress> addresses) {
-    void onStart(@Observes StartupEvent ev) {
-        if (executor == null) {
-            return;
+    void onStop(@Observes ShutdownEvent ev) {
```
I would use a `@PreDestroy` here instead.

vietk commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -189,56 +188,56 @@ private static String asString(List<InetSocketAddress> addresses) {
-        } else {
-            streams = new KafkaStreams(topology.get(), streamsProperties, kafkaClientSupplier.get());
+    @Produces
+    @Singleton
```
I checked on a small test, a Singleton bean seems to be not created if the bean is not injected anywhere.
Can someone confirms this behavior, I searched in CDI/Arc I did not find anything.
In our application we are not injecting (yet) a KafkaStreams instance, because we don't need it, so this is not gonna work.
Can you make sure it's working (or not) in that case : not injecting the bean somewhere.

vietk replied:
I found the doc : https://quarkus.io/guides/cdi-reference#lazy_by_default

` By default, CDI beans are created lazily, when needed.`
`A bean with a pseudo-scope (@Dependent and @Singleton ) is created when injected.`

vietk commented:
> A quick solution could be to inject KafkaStreams inside KafkaStreamsTopicsHealthCheck (even if not used) to make sure KafkaStreamsTopologyManager is full initialized ?

That would solve the concern I have about the creation of the bean

gsmet commented:
Yeah, we have race conditions. The admin config can be accessed even if not initialized properly. I think we need to properly fix that.

The main issue is how we push the config to the topology manager, especially the properties that are not Quarkus config. Sooner or later, this always ends up being an issue.

I'll take a look at it as it's going to be pretty hairy.

gsmet committed:
Reorganize the KafkaStreams beans initialization

This will avoid race conditions and make sure everything is properly
initialized before consuming the beans.

gsmet commented:
I pushed a commit with what I think should fix the issues.

Could you all have a look?
@geoand could you have a look too?

gsmet commented:
@mkouba if you could have a quick look too, that would be helpful

geoand commented:
The approach seems reasonable to me and inline with what we do in other extensions

gsmet commented:
Just a note that I would like to include this in 1.6.1.Final that I plan to release on Monday so if you people can find some time to check things with your app before Monday, that would be great.

pcasaes commented:
Just tested on a couple of projects. Everything OK.

gsmet commented:
Thanks @pcasaes !

Rebased with the latest fix from @mkouba and the `@Startup` annotations added again. Let's see what CI has to say.

geoand commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java:
```diff
@@ -0,0 +1,309 @@
+    private final KafkaStreams kafkaStreams;
+    private final KafkaStreamsTopologyManager kafkaStreamsTopologyManager;
+
+    @Inject
```
This annotation shouldn't be necessary

gsmet replied:
Yeah, I know it's not. It's purely cosmetic to make me think the parameters are injected :).

pcasaes commented:
Tested again and still good.

gunnarmorling commented:
Thanks all!

>

geoand commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -1,282 +1,29 @@
-    private volatile KafkaStreamsRuntimeConfig runtimeConfig;
-    private volatile Instance<Topology> topology;
-    private volatile Properties properties;
     private volatile Properties adminClientConfig;
```
This looks like it can be `final` now, correct?

gsmet replied:
Ah yeah, forgot to remove it.

gsmet replied:
Fixed. Removed `volatile` and added `final`.

geoand commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java:
```diff
@@ -0,0 +1,309 @@
+
+    @Produces
+    @Singleton
+    @Unremovable
```
Is it certain that this needs to be `@Unremovable`?

geoand commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java:
```diff
@@ -0,0 +1,309 @@
+
+    @Produces
+    @Singleton
+    @Unremovable
```
Same here

gsmet replied:
It was before so I mimic the existing behavior.

geoand reviewed (approved):
🎉

gsmet commented:
Thanks everyone!
